<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Queries the Human Cell Atlas • cellNexus</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Queries the Human Cell Atlas">
<meta property="og:description" content="Provides access to a copy of the Human Cell Atlas, but with harmonised metadata. This allows for uniform querying across numerous datasets within the Atlas using common fields such as cell type, tissue type, and patient ethnicity. Usage involves first querying the metadata table for cells of interest, and then downloading the corresponding cells into one of several supported formats, including SingleCellExperiment, Seurat, or pseudobulk SummarizedExperiment objects. Note: The Human Cell Atlas data accessed through this package is subject to its own licensing terms (typically Creative Commons Attribution or similar open data licenses as specified by the Human Cell Atlas Data Use Agreement), which differ from the package license (GPL-3). See the package documentation for details.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">cellNexus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/cellNexus.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MangiolaLaboratory/cellNexus/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="cellnexus">cellNexus<a class="anchor" aria-label="anchor" href="#cellnexus"></a>
</h1></div>
<!-- badges: start -->

</div>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p><code>cellNexus</code> builds upon and extends the functionality of the previously released <code>CuratedAtlasQueryR</code>, providing a unified query and access interface to the harmonised, curated, and reannotated CELLxGENE human cell atlas. It enables reproducible and programmatic exploration of large-scale single-cell data resources, supporting retrieval at the cell, sample, and dataset levels through flexible filtering by tissue, cell type, experimental condition, or other metadata features. The retrieved data are returned for downstream analysis.</p>
<p><code>cellNexus</code> integrates over 40 million human cells processed with standardised quality control, consistent normalisation, and unified abundance representations—including single-cell, counts-per-million, pseudobulk, and metacell layers. This harmonised design facilitates efficient cross-dataset analyses and downstream integration.</p>
<p>Data are hosted on the ARDC Nectar Research Cloud, and most <code>cellNexus</code> functions interact with Nectar via web requests, so a network connection is required for most functionality.</p>
<p><code>cellNexus</code> and CuratedAtlasQueryR both rely on pre-computed expression layers, but they differ in how these layers were generated. <code>cellNexus</code> applies a more standardised workflow with explicit empty droplet and dead cell removal followed by harmonised QC, normalisation, and multi-layer data generation. In doing so, it produces newly iterated data that align with the evolving CELLxGENE releases.</p>
<p><img src="reference/figures/logo.png" width="120x" height="139px"></p>
<p><img src="reference/figures/svcf_logo.jpeg" width="155x" height="58px"><img src="reference/figures/czi_logo.png" width="129px" height="58px"><img src="reference/figures/bioconductor_logo.jpg" width="202px" height="58px"><img src="reference/figures/vca_logo.png" width="219px" height="58px"><img src="reference/figures/nectar_logo.png" width="180px" height="58px"></p>
</div>
<div class="section level1">
<h1 id="query-interface">Query interface<a class="anchor" aria-label="anchor" href="#query-interface"></a>
</h1>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"MangiolaLaboratory/cellNexus"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-package">Load the package<a class="anchor" aria-label="anchor" href="#load-the-package"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MangiolaLaboratory/cellNexus" class="external-link">cellNexus</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-additional-packages">Load additional packages<a class="anchor" aria-label="anchor" href="#load-additional-packages"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-and-explore-the-metadata">Load and explore the metadata<a class="anchor" aria-label="anchor" href="#load-and-explore-the-metadata"></a>
</h2>
<div class="section level3">
<h3 id="load-the-metadata">Load the metadata<a class="anchor" aria-label="anchor" href="#load-the-metadata"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cloud_metadata <span class="op">=</span> <span class="va">METADATA_URL</span><span class="op">)</span></span>
<span><span class="va">metadata</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co">#&gt; ℹ Downloading 1 file, totalling 0 GB</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt; ℹ Downloading https://object-store.rc.nectar.org.au/v1/AUTH_06d6e008e3e642da99d806ba3ea629c5/cellNexus-metadata/sample_metadata.1.3.0.parquet to /tmp/RtmpVDjbPV/sample_metadata.1.3.0.parquet</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; # Source:   SQL [?? x 97]</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt; # Database: DuckDB 1.4.3 [unknown@Linux 5.14.0-362.24.1.el9_3.x86_64:R 4.5.2/:memory:]</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt;    cell_id               dataset_id observation_joinid sample_id cell_type cell_type_ontology_t…¹ sample_ assay</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;                 &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;                  &lt;chr&gt;   &lt;chr&gt;</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt;  1 FCAImmP7528294-ACATA… cda2c8cd-… p5e=WoIq0d         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt;  2 FCAImmP7528294-CATGA… cda2c8cd-… lx`7Bo-&amp;7n         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;  3 FCAImmP7528294-AAGAC… cda2c8cd-… *NUPW@J{c2         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt;  4 FCAImmP7528294-AATCG… cda2c8cd-… KIV&gt;qGFIS?         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt;  5 FCAImmP7528294-TTTGG… cda2c8cd-… *_#lQ&lt;oUnT         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt;  6 FCAImmP7528294-TCAGG… cda2c8cd-… zHCZWNmUHu         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt;  7 FCAImmP7528294-GACGC… cda2c8cd-… -NL-OH3!IA         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt;  8 FCAImmP7528294-CTTGG… cda2c8cd-… 6mRCZW}rOM         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt;  9 FCAImmP7528294-CACTC… cda2c8cd-… I6&gt;u{Gb-J_         034f0fb1… monocyte  CL:0000576             034f0f… 10x …</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt; 10 FCAImmP7579218-TGTGG… cda2c8cd-… IdHwp1GBZm         03ddfd57… monocyte  CL:0000576             03ddfd… 10x …</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; # ℹ more rows</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; # ℹ abbreviated name: ¹​cell_type_ontology_term_id</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; # ℹ 89 more variables: assay_ontology_term_id &lt;chr&gt;, cell_count &lt;int&gt;, citation &lt;chr&gt;, collection_id &lt;chr&gt;,</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co">#&gt; #   dataset_version_id &lt;chr&gt;, default_embedding &lt;chr&gt;, development_stage &lt;chr&gt;,</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="co">#&gt; #   development_stage_ontology_term_id &lt;chr&gt;, disease &lt;chr&gt;, disease_ontology_term_id &lt;chr&gt;, donor_id &lt;chr&gt;,</span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co">#&gt; #   experiment___ &lt;chr&gt;, explorer_url &lt;chr&gt;, feature_count &lt;int&gt;, filesize &lt;dbl&gt;, filetype &lt;chr&gt;,</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="co">#&gt; #   is_primary_data &lt;chr&gt;, mean_genes_per_cell &lt;dbl&gt;, organism &lt;chr&gt;, organism_ontology_term_id &lt;chr&gt;, …</span></span></code></pre></div>
<p>Metadata is saved to <code><a href="reference/get_default_cache_dir.html">get_default_cache_dir()</a></code> unless a custom path is provided via the cache_directory argument. The <code>metadata</code> variable can then be re-used for all subsequent queries.</p>
</div>
<div class="section level3">
<h3 id="explore-the-tissue">Explore the tissue<a class="anchor" aria-label="anchor" href="#explore-the-tissue"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="va">cell_type_unified_ensemble</span><span class="op">)</span> </span>
<span><span class="co">#&gt; # Source:   SQL [?? x 2]</span></span>
<span><span class="co">#&gt; # Database: DuckDB 1.4.3 [unknown@Linux 5.14.0-362.24.1.el9_3.x86_64:R 4.5.2/:memory:]</span></span>
<span><span class="co">#&gt;    tissue              cell_type_unified_ensemble</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;               &lt;chr&gt;                     </span></span>
<span><span class="co">#&gt;  1 thymus              cd14 mono                 </span></span>
<span><span class="co">#&gt;  2 breast              nk                        </span></span>
<span><span class="co">#&gt;  3 renal pelvis        epithelial                </span></span>
<span><span class="co">#&gt;  4 kidney              epithelial                </span></span>
<span><span class="co">#&gt;  5 kidney blood vessel epithelial                </span></span>
<span><span class="co">#&gt;  6 lung parenchyma     cd4 th2 em                </span></span>
<span><span class="co">#&gt;  7 respiratory airway  cd4 th2 em                </span></span>
<span><span class="co">#&gt;  8 lung                cd4 th1 em                </span></span>
<span><span class="co">#&gt;  9 lung                cd4 fh em                 </span></span>
<span><span class="co">#&gt; 10 nose                cd4 fh em                 </span></span>
<span><span class="co">#&gt; # ℹ more rows</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>cellNexus metadata applies standardised quality control to filter out empty droplets, dead or damaged cells, doublets, and samples with low gene counts.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">=</span> <span class="va">metadata</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">empty_droplet</span> <span class="op">==</span> <span class="cn">FALSE</span>,</span>
<span>         <span class="va">alive</span> <span class="op">==</span> <span class="cn">TRUE</span>,</span>
<span>         <span class="va">scDblFinder.class</span> <span class="op">!=</span> <span class="st">"doublet"</span>,</span>
<span>         <span class="va">feature_count</span> <span class="op">&gt;=</span> <span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="download-single-cell-rna-sequencing-counts">Download single-cell RNA sequencing counts<a class="anchor" aria-label="anchor" href="#download-single-cell-rna-sequencing-counts"></a>
</h2>
<div class="section level3">
<h3 id="query-raw-counts">Query raw counts<a class="anchor" aria-label="anchor" href="#query-raw-counts"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt; dim: 56239 6 </span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt; metadata(0):</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt; assays(1): counts</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt; rownames(56239): ENSG00000121410 ENSG00000268895 ... ENSG00000135605 ENSG00000109501</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; rowData names(0):</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt; colnames(6): LAP92_CATTCTAGTGCGGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt;   LAP92_CTCATGCCACCTGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1 ...</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt;   GCTCCTAAGGGTATCG_F02607___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;   AACACGTCACGCATCG_F01853___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; colData names(98): dataset_id observation_joinid ... dir_prefix original_cell_</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; reducedDimNames(0):</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt; mainExpName: NULL</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="query-counts-scaled-per-million">Query counts scaled per million<a class="anchor" aria-label="anchor" href="#query-counts-scaled-per-million"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_cpm</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_cpm</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#&gt; dim: 56239 6 </span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt; metadata(0):</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt; assays(1): cpm</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt; rownames(56239): ENSG00000121410 ENSG00000268895 ... ENSG00000135605 ENSG00000109501</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt; rowData names(0):</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt; colnames(6): LAP92_CATTCTAGTGCGGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt;   LAP92_CTCATGCCACCTGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1 ...</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt;   GCTCCTAAGGGTATCG_F02607___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt;   AACACGTCACGCATCG_F01853___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt; colData names(98): dataset_id observation_joinid ... dir_prefix original_cell_</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt; reducedDimNames(0):</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt; mainExpName: NULL</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="query-pseudobulk">Query pseudobulk<a class="anchor" aria-label="anchor" href="#query-pseudobulk"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudobulk_counts</span> <span class="op">=</span> </span>
<span>   <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/get_pseudobulk.html">get_pseudobulk</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pseudobulk_counts</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#&gt; dim: 56239 3 </span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co">#&gt; metadata(0):</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co">#&gt; assays(1): counts</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co">#&gt; rownames(56239): ENSG00000000003 ENSG00000000005 ... ENSG00000290292 ENSG00000291237</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co">#&gt; rowData names(0):</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co">#&gt; colnames(3): a2459ad4272363e6eb775e8e99607c3e___cd4 th1 em</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#&gt;   9c8fa5a8d2ae37179b579a0217670512___LAP92_1_duong___cd4 th2 em</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co">#&gt;   e4d7f8162faf68a85f61bdbd81dae627___cd4 th2 em</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="co">#&gt; colData names(59): dataset_id sample_id ... dir_prefix sample_identifier</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="co">#&gt; reducedDimNames(0):</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="co">#&gt; mainExpName: NULL</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="query-metacell">Query metacell<a class="anchor" aria-label="anchor" href="#query-metacell"></a>
</h3>
<p>The metadata includes a series of metacell aggregation levels, beginning with 2, 4, 8, and so on. For example, the value of metacell_2 represents a grouping of cells that can be split into two distinct metacells.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metacell_counts</span> <span class="op">=</span> </span>
<span>   <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">metacell_2</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/get_metacell.html">get_metacell</a></span><span class="op">(</span>cell_aggregation <span class="op">=</span> <span class="st">"metacell_2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metacell_counts</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co">#&gt; dim: 56239 4 </span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt; metadata(0):</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co">#&gt; assays(1): counts</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co">#&gt; rownames(56239): ENSG00000121410 ENSG00000268895 ... ENSG00000135605 ENSG00000109501</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt; rowData names(0):</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt; colnames(4): 9c8fa5a8d2ae37179b579a0217670512___LAP92_1_duong___1</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt;   9c8fa5a8d2ae37179b579a0217670512___LAP92_1_duong___2 e4d7f8162faf68a85f61bdbd81dae627___1</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt;   a2459ad4272363e6eb775e8e99607c3e___1</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt; colData names(39): metacell_2 dataset_id ... dir_prefix metacell_identifier</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt; reducedDimNames(0):</span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt; mainExpName: NULL</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-only-a-subset-of-genes">Extract only a subset of genes<a class="anchor" aria-label="anchor" href="#extract-only-a-subset-of-genes"></a>
</h3>
<p>This is helpful if just few genes are of interest (e.g ENSG00000134644 (PUM1)), as they can be compared across samples. cellNexus uses ENSEMBL gene ID(s).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_cpm</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span>, features <span class="op">=</span> <span class="st">"ENSG00000134644"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="co">#&gt; dim: 1 6 </span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co">#&gt; metadata(0):</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="co">#&gt; assays(1): cpm</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="co">#&gt; rownames(1): ENSG00000134644</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a><span class="co">#&gt; rowData names(0):</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co">#&gt; colnames(6): LAP92_CATTCTAGTGCGGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="co">#&gt;   LAP92_CTCATGCCACCTGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1 ...</span></span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a><span class="co">#&gt;   GCTCCTAAGGGTATCG_F02607___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a><span class="co">#&gt;   AACACGTCACGCATCG_F01853___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="co">#&gt; colData names(98): dataset_id observation_joinid ... dir_prefix original_cell_</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a><span class="co">#&gt; reducedDimNames(0):</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a><span class="co">#&gt; mainExpName: NULL</span></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-the-counts-as-a-seurat-object">Extract the counts as a Seurat object<a class="anchor" aria-label="anchor" href="#extract-the-counts-as-a-seurat-object"></a>
</h3>
<p>This convert the H5 SingleCellExperiment to Seurat so it might take long time and occupy a lot of memory depending on how many cells you are requesting.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/get_seurat.html">get_seurat</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_counts</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="co">#&gt; 56239 features across 6 samples within 1 assay </span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="co">#&gt; Active assay: originalexp (56239 features, 0 variable features)</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
<p>By default, data is downloaded to <code><a href="reference/get_default_cache_dir.html">get_default_cache_dir()</a></code> output. If memory is a concern, users can specify a custom cache directory to metadata and counts functions:</p>
</div>
</div>
<div class="section level2">
<h2 id="load-metadata-from-the-custom-cache-directory">Load metadata from the custom cache directory<a class="anchor" aria-label="anchor" href="#load-metadata-from-the-custom-cache-directory"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cache_directory <span class="op">=</span> <span class="st">"/MY/CUSTOM/PATH"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="query-raw-counts-from-the-custom-cache-directory">Query raw counts from the custom cache directory<a class="anchor" aria-label="anchor" href="#query-raw-counts-from-the-custom-cache-directory"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>cache_directory <span class="op">=</span> <span class="st">"/MY/CUSTOM/PATH"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre></div>
<p>Same strategy can be applied for functions <code>get_pseuodbulk()</code>, <code><a href="reference/get_metacell.html">get_metacell()</a></code>, <code><a href="reference/get_seurat.html">get_seurat()</a></code> by passing your custom directory character to “cache_directory” parameter.</p>
</div>
<div class="section level2">
<h2 id="save-your-singlecellexperiment">Save your <code>SingleCellExperiment</code>
<a class="anchor" aria-label="anchor" href="#save-your-singlecellexperiment"></a>
</h2>
<p>The returned <code>SingleCellExperiment</code> can be saved with three modalities, as <code>.rds</code> or as <code>HDF5</code> or as <code>H5AD</code>.</p>
<div class="section level3">
<h3 id="saving-as-rds-fast-saving-slow-reading">Saving as RDS (fast saving, slow reading)<a class="anchor" aria-label="anchor" href="#saving-as-rds-fast-saving-slow-reading"></a>
</h3>
<p>Saving as <code>.rds</code> has the advantage of being fast, and the <code>.rds</code> file occupies very little disk space as it only stores the links to the files in your cache.</p>
<p>However it has the disadvantage that for big <code>SingleCellExperiment</code> objects, which merge a lot of HDF5 from your <code>get_single_cell_experiment</code>, the display and manipulation is going to be slow. In addition, an <code>.rds</code> saved in this way is not portable: you will not be able to share it with other users.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="st">"single_cell_counts.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="saving-as-hdf5-slow-saving-fast-reading">Saving as HDF5 (slow saving, fast reading)<a class="anchor" aria-label="anchor" href="#saving-as-hdf5-slow-saving-fast-reading"></a>
</h3>
<p>Saving as <code>.hdf5</code> executes any computation on the <code>SingleCellExperiment</code> and writes it to disk as a monolithic <code>HDF5</code>. Once this is done, operations on the <code>SingleCellExperiment</code> will be comparatively very fast. The resulting <code>.hdf5</code> file will also be totally portable and sharable.</p>
<p>However this <code>.hdf5</code> has the disadvantage of being larger than the corresponding <code>.rds</code> as it includes a copy of the count information, and the saving process is going to be slow for large objects.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ! IMPORTANT if you save 200K+ cells</span></span>
<span><span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu">setAutoBlockSize</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1e+09</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html" class="external-link">saveHDF5SummarizedExperiment</a></span><span class="op">(</span></span>
<span>    <span class="st">"single_cell_counts"</span>, </span>
<span>    replace <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    as.sparse <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="saving-as-h5ad-slow-saving-fast-reading">Saving as H5AD (slow saving, fast reading)<a class="anchor" aria-label="anchor" href="#saving-as-h5ad-slow-saving-fast-reading"></a>
</h3>
<p>Saving as <code>.h5ad</code> executes any computation on the <code>SingleCellExperiment</code> and writes it to disk as a monolithic <code>H5AD</code>. The <code>H5AD</code> format is the HDF5 disk representation of the AnnData object and is well-supported in Python.</p>
<p>However this <code>.h5ad</code> saving strategy has a bottleneck of handling columns with only NA values of a <code>SingleCellExperiment</code> metadata.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ! IMPORTANT if you save 200K+ cells</span></span>
<span><span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu">setAutoBlockSize</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1e+09</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> <span class="fu">anndataR</span><span class="fu">::</span><span class="fu"><a href="https://anndataR.data-intuitive.com/reference/write_h5ad.html" class="external-link">write_h5ad</a></span><span class="op">(</span><span class="st">"single_cell_counts.h5ad"</span>, </span>
<span>                                               compression <span class="op">=</span> <span class="st">"gzip"</span>,</span>
<span>                                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualise-gene-transcription">Visualise gene transcription<a class="anchor" aria-label="anchor" href="#visualise-gene-transcription"></a>
</h2>
<p>We can gather all CD14 monocytes cells and plot the distribution of ENSG00000085265 (FCN1) across all tissues</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plots with styling</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># Filter and subset</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cell_type_unified_ensemble</span> <span class="op">==</span> <span class="st">"cd14 mono"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Get counts per million for FCN1 gene</span></span>
<span>  <span class="fu"><a href="reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span>, features <span class="op">=</span> <span class="st">"ENSG00000085265"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Add feature to table</span></span>
<span>  <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu">join_features</span><span class="op">(</span><span class="st">"ENSG00000085265"</span>, shape <span class="op">=</span> <span class="st">"wide"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Rank x axis</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Rename to gene symbol</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>FCN1 <span class="op">=</span> <span class="va">ENSG00000085265</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot by disease</span></span>
<span><span class="va">counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/with_groups.html" class="external-link">with_groups</a></span><span class="op">(</span><span class="va">disease</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">`FCN1`</span>, rm.na<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html" class="external-link">fct_reorder</a></span><span class="op">(</span><span class="va">disease</span>, <span class="va">median_count</span>,.desc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">`FCN1`</span>,color <span class="op">=</span> <span class="va">dataset_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>shape<span class="op">=</span><span class="st">"."</span><span class="op">)</span> <span class="op">+</span></span>
<span>    </span>
<span>  <span class="co"># Style</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">60</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Disease"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"FCN1 in CD14 monocytes by disease. Coloured by datasets"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="reference/figures/FCN1_disease_plot.png"><!-- --></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot by tissue</span></span>
<span><span class="va">counts</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/with_groups.html" class="external-link">with_groups</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">`FCN1`</span>, rm.na<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html" class="external-link">fct_reorder</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="va">median_count</span>,.desc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">`FCN1`</span>,color <span class="op">=</span> <span class="va">dataset_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>shape<span class="op">=</span><span class="st">"."</span><span class="op">)</span> <span class="op">+</span></span>
<span>    </span>
<span>  <span class="co"># Style</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">60</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Tissue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"FCN1 in CD14 monocytes by tissue. Colored by datasets"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/FCN1_tissue_plot.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="integrate-cloud-and-local-metadata">Integrate cloud and local metadata<a class="anchor" aria-label="anchor" href="#integrate-cloud-and-local-metadata"></a>
</h2>
<p><code>cellNexus</code> not only enables users to query our metadata but also allows integration with your local metadata. Additionally, users can integrate with your metadata stored in the cloud.</p>
<p>To enable this feature, users must include <code>file_id_cellNexus_single_cell</code> and <code>atlas_id</code> (e.g cellxgene/dd-mm-yy) columns in the metadata. See metadata structure in cellNexus::pbmc3k_sce</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up local cache and paths</span></span>
<span><span class="va">local_cache</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">layer</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span><span class="va">meta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">local_cache</span>, <span class="st">"pbmc3k_metadata.parquet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract and prepare metadata</span></span>
<span><span class="va">pbmc3k_metadata</span> <span class="op">&lt;-</span> <span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="reference/pbmc3k_sce.html">pbmc3k_sce</a></span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    counts_directory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">atlas_id</span>, <span class="va">layer</span><span class="op">)</span>,</span>
<span>    sce_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">counts_directory</span>, <span class="va">file_id_cellNexus_single_cell</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Get unique paths</span></span>
<span><span class="va">counts_directory</span> <span class="op">&lt;-</span> <span class="va">pbmc3k_metadata</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">counts_directory</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_path</span> <span class="op">&lt;-</span> <span class="va">pbmc3k_metadata</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">sce_path</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create directory structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">counts_directory</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save data to disk</span></span>
<span><span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="reference/pbmc3k_sce.html">pbmc3k_sce</a></span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/write_parquet.html" class="external-link">write_parquet</a></span><span class="op">(</span><span class="va">meta_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save SCE object</span></span>
<span><span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="reference/pbmc3k_sce.html">pbmc3k_sce</a></span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">anndataR</span><span class="fu">::</span><span class="fu"><a href="https://anndataR.data-intuitive.com/reference/write_h5ad.html" class="external-link">write_h5ad</a></span><span class="op">(</span><span class="va">sce_path</span>, compression <span class="op">=</span> <span class="st">"gzip"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A cellNexus file</span></span>
<span><span class="va">file_id_from_cloud</span> <span class="op">&lt;-</span> <span class="st">"e52795dec7b626b6276b867d55328d9f___1.h5ad"</span></span>
<span><span class="va">file_id_local</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">sce_path</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cloud_metadata <span class="op">=</span> <span class="va">METADATA_URL</span>,</span>
<span>             local_metadata <span class="op">=</span> <span class="va">meta_path</span>,</span>
<span>             cache_directory <span class="op">=</span> <span class="va">local_cache</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># For illustration purpose, only filter a selected cloud metadata and the saved metadata </span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">file_id_cellNexus_single_cell</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">file_id_from_cloud</span>, <span class="va">file_id_local</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_id</span>, <span class="va">sample_id</span>, <span class="va">dataset_id</span>, <span class="va">cell_type_unified_ensemble</span>, <span class="va">atlas_id</span>, <span class="va">file_id_cellNexus_single_cell</span> <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>cache_directory <span class="op">=</span> <span class="va">local_cache</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Realising metadata.</span></span>
<span><span class="co">#&gt; ℹ Synchronising files</span></span>
<span><span class="co">#&gt; ℹ Downloading 1 file, totalling 0.02 GB</span></span>
<span><span class="co">#&gt; ℹ Downloading https://object-store.rc.nectar.org.au/v1/AUTH_06d6e008e3e642da99d806ba3ea629c5/cellNexus-anndata/cellxgene/21-08-2025/counts/e52795dec7b626b6276b867d55328d9f___1.h5ad to /tmp/RtmpVDjbPV/cellxgene/21-08-2025//counts/e52795dec7b626b6276b867d55328d9f___1.h5ad</span></span>
<span><span class="co">#&gt; ℹ Reading files.</span></span>
<span><span class="co">#&gt; ! cellNexus says: Not all genes completely overlap across the provided objects.Counts are generated by genes intersection.</span></span>
<span><span class="co">#&gt; ℹ Compiling Experiment.</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 12795 3572 </span></span>
<span><span class="co">#&gt; metadata(1): data</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(12795): ENSG00000228463 ENSG00000228327 ... ENSG00000273748 ENSG00000278384</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(3572): AAACATACAACCAC_1 AAACATTGAGCTAC_1 ...</span></span>
<span><span class="co">#&gt;   TCACAAGAGTTGAGTA_5_liao___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span><span class="co">#&gt;   AGGGTGACACGCATCG_5_liao___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span><span class="co">#&gt; colData names(7): sample_id dataset_id ... dir_prefix original_cell_</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="cell-metadata">Cell metadata<a class="anchor" aria-label="anchor" href="#cell-metadata"></a>
</h1>
<p>Dataset-specific columns (definitions available at cellxgene.cziscience.com)</p>
<p><code>cell_count</code>, <code>collection_id</code>, <code>filetype</code>, <code>is_primary_data</code>, <code>mean_genes_per_cell</code>, <code>published_at</code>, <code>revised_at</code>, <code>schema_version</code>, <code>tombstone</code>, <code>x_normalization</code>, <code>explorer_url</code>, <code>dataset_id</code>, <code>dataset_version_id</code></p>
<p>Sample-specific columns (definitions available at cellxgene.cziscience.com)</p>
<p><code>sample_id</code>, <code>sample_</code>, <code>age_days</code>, <code>assay</code>, <code>assay_ontology_term_id</code>, <code>development_stage</code>, <code>development_stage_ontology_term_id</code>, <code>self_reported_ethnicity</code>, <code>self_reported_ethnicity_ontology_term_id</code>, <code>experiment___</code>, <code>organism</code>, <code>organism_ontology_term_id</code>, <code>sample_placeholder</code>, <code>sex</code>, <code>sex_ontology_term_id</code>, <code>tissue</code>, <code>tissue_type</code>, <code>tissue_ontology_term_id</code>, <code>tissue_groups</code>, <code>disease</code>, <code>disease_ontology_term_id</code>, <code>is_primary_data</code>, <code>donor_id</code>, <code>is_immune</code></p>
<p>Cell-specific columns (definitions available at cellxgene.cziscience.com)</p>
<p><code>cell_id</code>, <code>cell_type</code>, <code>cell_type_ontology_term_id</code>, <code>cell_annotation_azimuth_l2</code>, <code>cell_annotation_blueprint_singler</code>, <code>observation_joinid</code>, <code>empty_droplet</code>, <code>alive</code>, <code>scDblFinder.class</code></p>
<p>Through harmonisation and curation we introduced custom column, not present in the original CELLxGENE metadata</p>
<ul>
<li>
<code>age_days</code>: donors’ age in days</li>
<li>
<code>cell_type_unified_ensemble</code>: the consensus call identity (for immune cells) using the original and three novel annotations using Seurat Azimuth and SingleR</li>
<li>
<code>cell_annotation_azimuth_l2</code>: Azimuth cell annotation</li>
<li>
<code>cell_annotation_blueprint_singler</code>: SingleR cell annotation using Blueprint reference</li>
<li>
<code>cell_annotation_blueprint_monaco</code>: SingleR cell annotation using Monaco reference</li>
<li>
<code>sample_heuristic</code>: sample subdivision for internal use</li>
<li>
<code>file_id_cellNexus_single_cell</code>: file subdivision for internal use</li>
<li>
<code>file_id_cellNexus_pseudobulk</code>: file subdivision for internal use</li>
<li>
<code>sample_id</code>: sample ID</li>
<li>
<code>nCount_RNA</code>: total number of RNA detected in a cell per sample</li>
<li>
<code>nFeature_expressed_in_sample</code>: total number of genes expressed in a cell per sample</li>
</ul>
</div>
<div class="section level1">
<h1 id="rna-abundance">RNA abundance<a class="anchor" aria-label="anchor" href="#rna-abundance"></a>
</h1>
<p>The <code>counts</code> assay includes RNA abundance in the positive real scale (not transformed with non-linear functions, e.g. log sqrt). Originally CELLxGENE include a mix of scales and transformations specified in the <code>x_normalization</code> column.</p>
<p>The <code>cpm</code> assay includes counts per million.</p>
</div>
<div class="section level1">
<h1 id="other-representations">Other representations<a class="anchor" aria-label="anchor" href="#other-representations"></a>
</h1>
<p>The <code>rank</code> assay is the representation of each cell’s gene expression profile where genes are ranked by expression intensity.</p>
<p>The <code>pseudobulk</code> assay includes aggregated RNA abundance for sample and cell type combination.</p>
<p>The metacell (e.g <code>metacell_2</code>, <code>metacell_4</code> etc) assays represent hierarchical partitions of cells into metacell groups.</p>
</div>
<div class="section level1">
<h1 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Red Hat Enterprise Linux 9.3 (Plow)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /stornext/System/data/software/rhel/9/base/tools/R/4.5.2/lib64/R/lib/libRblas.so </span></span>
<span><span class="co">#&gt; LAPACK: /stornext/System/data/software/rhel/9/base/tools/R/4.5.2/lib64/R/lib/libRlapack.so;  LAPACK version 3.12.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Australia/Melbourne</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_4.0.1               cellNexus_0.99.5            stringr_1.6.0              </span></span>
<span><span class="co">#&gt;  [4] dplyr_1.1.4                 SummarizedExperiment_1.40.0 Biobase_2.70.0             </span></span>
<span><span class="co">#&gt;  [7] GenomicRanges_1.62.1        Seqinfo_1.0.0               IRanges_2.44.0             </span></span>
<span><span class="co">#&gt; [10] S4Vectors_0.48.0            BiocGenerics_0.56.0         generics_0.1.4             </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.22.0       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22            splines_4.5.2               later_1.4.4                </span></span>
<span><span class="co">#&gt;   [4] filelock_1.0.3              tibble_3.3.0                polyclip_1.10-7            </span></span>
<span><span class="co">#&gt;   [7] fastDummies_1.7.5           lifecycle_1.0.4             rprojroot_2.1.1            </span></span>
<span><span class="co">#&gt;  [10] globals_0.18.0              lattice_0.22-7              MASS_7.3-65                </span></span>
<span><span class="co">#&gt;  [13] backports_1.5.0             magrittr_2.0.4              plotly_4.11.0              </span></span>
<span><span class="co">#&gt;  [16] sass_0.4.10                 rmarkdown_2.30              remotes_2.5.0              </span></span>
<span><span class="co">#&gt;  [19] jquerylib_0.1.4             yaml_2.3.12                 httpuv_1.6.16              </span></span>
<span><span class="co">#&gt;  [22] otel_0.2.0                  Seurat_5.4.0                sctransform_0.4.2          </span></span>
<span><span class="co">#&gt;  [25] spam_2.11-1                 sessioninfo_1.2.3           pkgbuild_1.4.8             </span></span>
<span><span class="co">#&gt;  [28] sp_2.2-0                    spatstat.sparse_3.1-0       reticulate_1.44.1          </span></span>
<span><span class="co">#&gt;  [31] cowplot_1.2.0               pbapply_1.7-4               DBI_1.2.3                  </span></span>
<span><span class="co">#&gt;  [34] RColorBrewer_1.1-3          pkgload_1.4.1               abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [37] Rtsne_0.17                  purrr_1.2.0                 ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [40] irlba_2.3.5.1               listenv_0.10.0              spatstat.utils_3.2-0       </span></span>
<span><span class="co">#&gt;  [43] testthat_3.3.1              goftest_1.2-3               RSpectra_0.16-2            </span></span>
<span><span class="co">#&gt;  [46] spatstat.random_3.4-3       fitdistrplus_1.2-4          parallelly_1.46.0          </span></span>
<span><span class="co">#&gt;  [49] commonmark_2.0.0            codetools_0.2-20            DelayedArray_0.36.0        </span></span>
<span><span class="co">#&gt;  [52] xml2_1.5.1                  tidyselect_1.2.1            UCSC.utils_1.6.0           </span></span>
<span><span class="co">#&gt;  [55] farver_2.1.2                shinyWidgets_0.9.0          spatstat.explore_3.6-0     </span></span>
<span><span class="co">#&gt;  [58] duckdb_1.4.3                roxygen2_7.3.3              jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt;  [61] ellipsis_0.3.2              progressr_0.18.0            ggridges_0.5.7             </span></span>
<span><span class="co">#&gt;  [64] survival_3.8-3              tools_4.5.2                 ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [67] Rcpp_1.1.0                  glue_1.8.0                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [70] SparseArray_1.10.6          xfun_0.54                   usethis_3.2.1              </span></span>
<span><span class="co">#&gt;  [73] GenomeInfoDb_1.46.2         HDF5Array_1.38.0            withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [76] fastmap_1.2.0               basilisk_1.22.0             rhdf5filters_1.22.0        </span></span>
<span><span class="co">#&gt;  [79] digest_0.6.39               R6_2.6.1                    mime_0.13                  </span></span>
<span><span class="co">#&gt;  [82] scattermore_1.2             tensor_1.5.1                spatstat.data_3.1-9        </span></span>
<span><span class="co">#&gt;  [85] h5mread_1.2.1               utf8_1.2.6                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [88] data.table_1.17.8           httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt;  [91] S4Arrays_1.10.1             uwot_0.2.4                  pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [94] gtable_0.3.6                blob_1.2.4                  lmtest_0.9-40              </span></span>
<span><span class="co">#&gt;  [97] S7_0.2.1                    SingleCellExperiment_1.32.0 XVector_0.50.0             </span></span>
<span><span class="co">#&gt; [100] brio_1.1.5                  htmltools_0.5.9             dotCall64_1.2              </span></span>
<span><span class="co">#&gt; [103] SeuratObject_5.3.0          scales_1.4.0                png_0.1-8                  </span></span>
<span><span class="co">#&gt; [106] spatstat.univar_3.1-5       knitr_1.50                  rstudioapi_0.17.1          </span></span>
<span><span class="co">#&gt; [109] reshape2_1.4.5              checkmate_2.3.3             nlme_3.1-168               </span></span>
<span><span class="co">#&gt; [112] curl_7.0.0                  anndataR_1.0.0              cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [115] zoo_1.8-14                  rhdf5_2.54.1                KernSmooth_2.23-26         </span></span>
<span><span class="co">#&gt; [118] parallel_4.5.2              miniUI_0.1.2                arrow_22.0.0               </span></span>
<span><span class="co">#&gt; [121] zellkonverter_1.20.0        desc_1.4.3                  pillar_1.11.1              </span></span>
<span><span class="co">#&gt; [124] grid_4.5.2                  vctrs_0.6.5                 RANN_2.6.2                 </span></span>
<span><span class="co">#&gt; [127] promises_1.5.0              dbplyr_2.5.1                xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [130] cluster_2.1.8.1             evaluate_1.0.5              cli_3.6.5                  </span></span>
<span><span class="co">#&gt; [133] compiler_4.5.2              rlang_1.1.6                 future.apply_1.20.1        </span></span>
<span><span class="co">#&gt; [136] plyr_1.8.9                  fs_1.6.6                    stringi_1.8.7              </span></span>
<span><span class="co">#&gt; [139] viridisLite_0.4.2           deldir_2.0-4                assertthat_0.2.1           </span></span>
<span><span class="co">#&gt; [142] lazyeval_0.2.2              devtools_2.4.6              spatstat.geom_3.6-1        </span></span>
<span><span class="co">#&gt; [145] Matrix_1.7-4                dir.expiry_1.18.0           RcppHNSW_0.6.0             </span></span>
<span><span class="co">#&gt; [148] patchwork_1.3.2             bit64_4.6.0-1               future_1.68.0              </span></span>
<span><span class="co">#&gt; [151] Rhdf5lib_1.32.0             shiny_1.12.1                ROCR_1.0-11                </span></span>
<span><span class="co">#&gt; [154] igraph_2.2.1                memoise_2.0.1               bslib_0.9.0                </span></span>
<span><span class="co">#&gt; [157] bit_4.6.0</span></span></code></pre></div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MangiolaLaboratory/cellNexus/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MangiolaLaboratory/cellNexus/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing cellNexus</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, reviewer, contributor </small> <a href="https://orcid.org/0000-0001-7474-836X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Mengyuan Shen <br><small class="roles"> Author, maintainer, reviewer </small> <a href="https://orcid.org/0009-0008-0057-8239" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Michael Milton <br><small class="roles"> Author, reviewer, contributor </small> <a href="https://orcid.org/0000-0002-8965-2595" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Silicon Valley Foundation CZF2019-002443 <br><small class="roles"> Funder </small>   </li>
<li>NIH NHGRI 5U24HG004059-18 <br><small class="roles"> Funder </small>   </li>
<li>Victoria Cancer Agency ECRF21036 <br><small class="roles"> Funder </small>   </li>
<li>NHMRC 1116955 <br><small class="roles"> Funder </small>   </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola, Mengyuan Shen, Michael Milton, Silicon Valley Foundation CZF2019-002443, NIH NHGRI 5U24HG004059-18, Victoria Cancer Agency ECRF21036, NHMRC 1116955.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
