<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>cellNexus • cellNexus</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="cellNexus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cellNexus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">cellNexus</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MangiolaLaboratory/cellNexus/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>cellNexus</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MangiolaLaboratory/cellNexus/blob/master/vignettes/Introduction.Rmd" class="external-link"><code>vignettes/Introduction.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <!-- badges: end --></p>
<p><code>cellNexus</code> is a query interface that allow the
programmatic exploration and retrieval of the harmonised, curated and
reannotated CELLxGENE single-cell human cell atlas. Data can be
retrieved at cell, sample, or dataset levels based on filtering
criteria.</p>
<p>Harmonised data is stored in the ARDC Nectar Research Cloud, and most
<code>cellNexus</code> functions interact with Nectar via web requests,
so a network connection is required for most functionality.</p>
<p><img src="../reference/figures/logo.png" width="120x" height="139px"></p>
<p><img src="../reference/figures/svcf_logo.jpeg" width="155x" height="58px"><img src="../reference/figures/czi_logo.png" width="129px" height="58px"><img src="../reference/figures/bioconductor_logo.jpg" width="202px" height="58px"><img src="../reference/figures/vca_logo.png" width="219px" height="58px"><img src="../reference/figures/nectar_logo.png" width="180px" height="58px"></p>
<div class="section level2">
<h2 id="query-interface">Query interface<a class="anchor" aria-label="anchor" href="#query-interface"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"MangiolaLaboratory/cellNexus"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-the-package">Load the package<a class="anchor" aria-label="anchor" href="#load-the-package"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MangiolaLaboratory/cellNexus" class="external-link">cellNexus</a></span><span class="op">)</span></span></code></pre></div>
<p>Load additional packages</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-and-explore-the-metadata">Load and explore the metadata<a class="anchor" aria-label="anchor" href="#load-and-explore-the-metadata"></a>
</h3>
<div class="section level4">
<h4 id="load-the-metadata">Load the metadata<a class="anchor" aria-label="anchor" href="#load-the-metadata"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cloud_metadata <span class="op">=</span> <span class="va">METADATA_URL</span><span class="op">)</span></span>
<span><span class="va">metadata</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Downloading 1 file, totalling 0 GB</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Downloading https://object-store.rc.nectar.org.au/v1/AUTH_06d6e008e3e642da99d806ba3ea629c5/cellNexus-metadata/sample_metadata.1.2.13.parquet to /tmp/RtmprGqjm3/sample_metadata.1.2.13.parquet</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:   SQL [?? x 94]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database: DuckDB 1.4.0 [unknown@Linux 6.11.0-1018-azure:R 4.5.1/:memory:]</span></span></span>
<span><span class="co">#&gt;    cell_id                     dataset_id observation_joinid sample_id cell_type</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> FCAImmP7528294-ACATACGAGTA… cda2c8cd-… p5e=WoIq0d         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> FCAImmP7528294-CACTCCACACC… cda2c8cd-… I6&gt;u{Gb-J_         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> FCAImmP7528294-TTTGGTTGTGG… cda2c8cd-… *_#lQ&lt;oUnT         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> FCAImmP7528294-TCAGGTACAAT… cda2c8cd-… zHCZWNmUHu         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> FCAImmP7528294-GACGCGTAGGC… cda2c8cd-… -NL-OH3!IA         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> FCAImmP7528294-CTTGGCTGTTC… cda2c8cd-… 6mRCZW}rOM         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> FCAImmP7528294-CATGACAGTCC… cda2c8cd-… lx`7Bo-&amp;7n         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> FCAImmP7528294-AAGACCTAGCT… cda2c8cd-… *NUPW@J{c2         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> FCAImmP7528294-AATCGGTTCAA… cda2c8cd-… KIV&gt;qGFIS?         034f0fb1… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> FCAImmP7579218-TGTGGTAGTCT… cda2c8cd-… IdHwp1GBZm         03ddfd57… monocyte </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 89 more variables: cell_type_ontology_term_id &lt;chr&gt;, sample_ &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   assay &lt;chr&gt;, assay_ontology_term_id &lt;chr&gt;, cell_count &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   citation &lt;chr&gt;, collection_id &lt;chr&gt;, dataset_version_id &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   default_embedding &lt;chr&gt;, development_stage &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   development_stage_ontology_term_id &lt;chr&gt;, disease &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   disease_ontology_term_id &lt;chr&gt;, donor_id &lt;chr&gt;, experiment___ &lt;chr&gt;, …</span></span></span></code></pre>
<p>Metadata is saved to <code><a href="../reference/get_default_cache_dir.html">get_default_cache_dir()</a></code> unless a
custom path is provided via the cache_directory argument. The
<code>metadata</code> variable can then be re-used for all subsequent
queries.</p>
</div>
<div class="section level4">
<h4 id="explore-the-tissue">Explore the tissue<a class="anchor" aria-label="anchor" href="#explore-the-tissue"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="va">cell_type_unified_ensemble</span><span class="op">)</span> </span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:   SQL [?? x 2]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database: DuckDB 1.4.0 [unknown@Linux 6.11.0-1018-azure:R 4.5.1/:memory:]</span></span></span>
<span><span class="co">#&gt;    tissue             cell_type_unified_ensemble</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> thymus             cd14 mono                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> breast             nk                        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> renal pelvis       epithelial                </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> respiratory airway cd4 th2 em                </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> lung               cd4 th1 em                </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> nose               cd4 fh em                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> respiratory airway granulocyte               </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> blood              cd14 mono                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> nose               cd4 tcm                   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> colonic epithelium nk                        </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ more rows</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>cellNexus metadata applies standardised quality control to filter out
empty droplets, dead or damaged cells, doublets, and samples with low
gene counts.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">=</span> <span class="va">metadata</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">empty_droplet</span> <span class="op">==</span> <span class="cn">FALSE</span>,</span>
<span>         <span class="va">alive</span> <span class="op">==</span> <span class="cn">TRUE</span>,</span>
<span>         <span class="va">scDblFinder.class</span> <span class="op">!=</span> <span class="st">"doublet"</span>,</span>
<span>         <span class="va">feature_count</span> <span class="op">&gt;=</span> <span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="download-single-cell-rna-sequencing-counts">Download single-cell RNA sequencing counts<a class="anchor" aria-label="anchor" href="#download-single-cell-rna-sequencing-counts"></a>
</h3>
<div class="section level4">
<h4 id="query-raw-counts">Query raw counts<a class="anchor" aria-label="anchor" href="#query-raw-counts"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 56239 1055 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(56239): ENSG00000121410 ENSG00000268895 ... ENSG00000135605</span></span>
<span><span class="co">#&gt;   ENSG00000109501</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1055):</span></span>
<span><span class="co">#&gt;   LAP92_CATTCTAGTGCGGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span><span class="co">#&gt;   LAP92_CTCATGCCACCTGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span><span class="co">#&gt;   ...</span></span>
<span><span class="co">#&gt;   LAP87_ACTTCGCCAAGACAAT-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_51</span></span>
<span><span class="co">#&gt;   LAP87_CGTCAAAAGACGGAAA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_51</span></span>
<span><span class="co">#&gt; colData names(95): dataset_id observation_joinid ... dir_prefix</span></span>
<span><span class="co">#&gt;   original_cell_</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="query-counts-scaled-per-million">Query counts scaled per million<a class="anchor" aria-label="anchor" href="#query-counts-scaled-per-million"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 56239 1055 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): cpm</span></span>
<span><span class="co">#&gt; rownames(56239): ENSG00000121410 ENSG00000268895 ... ENSG00000135605</span></span>
<span><span class="co">#&gt;   ENSG00000109501</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1055):</span></span>
<span><span class="co">#&gt;   LAP92_CATTCTAGTGCGGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span><span class="co">#&gt;   LAP92_CTCATGCCACCTGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span><span class="co">#&gt;   ...</span></span>
<span><span class="co">#&gt;   LAP87_ACTTCGCCAAGACAAT-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_51</span></span>
<span><span class="co">#&gt;   LAP87_CGTCAAAAGACGGAAA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_51</span></span>
<span><span class="co">#&gt; colData names(95): dataset_id observation_joinid ... dir_prefix</span></span>
<span><span class="co">#&gt;   original_cell_</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="query-pseudobulk">Query pseudobulk<a class="anchor" aria-label="anchor" href="#query-pseudobulk"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudobulk_counts</span> <span class="op">=</span> </span>
<span>   <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_pseudobulk.html">get_pseudobulk</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span></span>
<span><span class="va">pseudobulk_counts</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 56239 68 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(56239): ENSG00000000003 ENSG00000000005 ... ENSG00000290292</span></span>
<span><span class="co">#&gt;   ENSG00000291237</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(68): a2459ad4272363e6eb775e8e99607c3e___cd4 th1 em</span></span>
<span><span class="co">#&gt;   bfe624d44f7e5868cc22e11ad0f13866___cd4 th1 em ...</span></span>
<span><span class="co">#&gt;   270eb221dd0456cc063240404aec74cd___cd4 th1/th17 em</span></span>
<span><span class="co">#&gt;   9c8fa5a8d2ae37179b579a0217670512___LAP87_1_duong___cd4 th2 em</span></span>
<span><span class="co">#&gt; colData names(56): dataset_id sample_id ... dir_prefix</span></span>
<span><span class="co">#&gt;   sample_identifier</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="query-metacell">Query metacell<a class="anchor" aria-label="anchor" href="#query-metacell"></a>
</h4>
<p>The metadata includes a series of metacell aggregation levels,
beginning with 2, 4, 8, and so on. For example, the value of metacell_2
represents a grouping of cells that can be split into two distinct
metacells.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metacell_counts</span> <span class="op">=</span> </span>
<span>   <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">metacell_2</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_metacell.html">get_metacell</a></span><span class="op">(</span>cell_aggregation <span class="op">=</span> <span class="st">"metacell_2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span></span>
<span><span class="va">metacell_counts</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 56239 543 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(56239): ENSG00000121410 ENSG00000268895 ... ENSG00000135605</span></span>
<span><span class="co">#&gt;   ENSG00000109501</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(543): 9c8fa5a8d2ae37179b579a0217670512___LAP92_1_duong___1</span></span>
<span><span class="co">#&gt;   9c8fa5a8d2ae37179b579a0217670512___LAP92_1_duong___2 ...</span></span>
<span><span class="co">#&gt;   9c8fa5a8d2ae37179b579a0217670512___LAP87_1_duong___1</span></span>
<span><span class="co">#&gt;   9c8fa5a8d2ae37179b579a0217670512___LAP87_1_duong___2</span></span>
<span><span class="co">#&gt; colData names(39): metacell_2 dataset_id ... dir_prefix</span></span>
<span><span class="co">#&gt;   metacell_identifier</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-only-a-subset-of-genes">Extract only a subset of genes<a class="anchor" aria-label="anchor" href="#extract-only-a-subset-of-genes"></a>
</h4>
<p>This is helpful if just few genes are of interest (e.g
ENSG00000134644 (PUM1)), as they can be compared across samples.
cellNexus uses ENSEMBL gene ID(s).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span>, features <span class="op">=</span> <span class="st">"ENSG00000134644"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 1 1055 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): cpm</span></span>
<span><span class="co">#&gt; rownames(1): ENSG00000134644</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1055):</span></span>
<span><span class="co">#&gt;   LAP92_CATTCTAGTGCGGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span><span class="co">#&gt;   LAP92_CTCATGCCACCTGATA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_1</span></span>
<span><span class="co">#&gt;   ...</span></span>
<span><span class="co">#&gt;   LAP87_ACTTCGCCAAGACAAT-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_51</span></span>
<span><span class="co">#&gt;   LAP87_CGTCAAAAGACGGAAA-1_duong___9f222629-9e39-47d0-b83f-e08d610c7479_51</span></span>
<span><span class="co">#&gt; colData names(95): dataset_id observation_joinid ... dir_prefix</span></span>
<span><span class="co">#&gt;   original_cell_</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-the-counts-as-a-seurat-object">Extract the counts as a Seurat object<a class="anchor" aria-label="anchor" href="#extract-the-counts-as-a-seurat-object"></a>
</h4>
<p>This convert the H5 SingleCellExperiment to Seurat so it might take
long time and occupy a lot of memory depending on how many cells you are
requesting.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_seurat.html">get_seurat</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span><span class="co">#&gt; Warning: `PackageCheck()` was deprecated in SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `rlang::check_installed()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span><span class="va">seurat_counts</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 56239 features across 1055 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: originalexp (56239 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
<p>By default, data is downloaded to
<code><a href="../reference/get_default_cache_dir.html">get_default_cache_dir()</a></code> output. If memory is a concern,
users can specify a custom cache directory to metadata and counts
functions:</p>
</div>
</div>
<div class="section level3">
<h3 id="load-metadata-from-the-custom-cache-directory">Load metadata from the custom cache directory<a class="anchor" aria-label="anchor" href="#load-metadata-from-the-custom-cache-directory"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cache_directory <span class="op">=</span> <span class="st">"/MY/CUSTOM/PATH"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="query-raw-counts-from-the-custom-cache-directory">Query raw counts from the custom cache directory<a class="anchor" aria-label="anchor" href="#query-raw-counts-from-the-custom-cache-directory"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">=</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">self_reported_ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">assay</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">cell_type</span> <span class="op">|&gt;</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>cache_directory <span class="op">=</span> <span class="st">"/MY/CUSTOM/PATH"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre></div>
<p>Same strategy can be applied for functions
<code>get_pseuodbulk()</code>, <code><a href="../reference/get_metacell.html">get_metacell()</a></code>,
<code><a href="../reference/get_seurat.html">get_seurat()</a></code> by passing your custom directory character to
“cache_directory” parameter.</p>
</div>
<div class="section level3">
<h3 id="save-your-singlecellexperiment">Save your <code>SingleCellExperiment</code><a class="anchor" aria-label="anchor" href="#save-your-singlecellexperiment"></a>
</h3>
<p>The returned <code>SingleCellExperiment</code> can be saved with
three modalities, as <code>.rds</code> or as <code>HDF5</code> or as
<code>H5AD</code>.</p>
<div class="section level4">
<h4 id="saving-as-rds-fast-saving-slow-reading">Saving as RDS (fast saving, slow reading)<a class="anchor" aria-label="anchor" href="#saving-as-rds-fast-saving-slow-reading"></a>
</h4>
<p>Saving as <code>.rds</code> has the advantage of being fast, and the
<code>.rds</code> file occupies very little disk space as it only stores
the links to the files in your cache.</p>
<p>However it has the disadvantage that for big
<code>SingleCellExperiment</code> objects, which merge a lot of HDF5
from your <code>get_single_cell_experiment</code>, the display and
manipulation is going to be slow. In addition, an <code>.rds</code>
saved in this way is not portable: you will not be able to share it with
other users.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="st">"single_cell_counts.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="saving-as-hdf5-slow-saving-fast-reading">Saving as HDF5 (slow saving, fast reading)<a class="anchor" aria-label="anchor" href="#saving-as-hdf5-slow-saving-fast-reading"></a>
</h4>
<p>Saving as <code>.hdf5</code> executes any computation on the
<code>SingleCellExperiment</code> and writes it to disk as a monolithic
<code>HDF5</code>. Once this is done, operations on the
<code>SingleCellExperiment</code> will be comparatively very fast. The
resulting <code>.hdf5</code> file will also be totally portable and
sharable.</p>
<p>However this <code>.hdf5</code> has the disadvantage of being larger
than the corresponding <code>.rds</code> as it includes a copy of the
count information, and the saving process is going to be slow for large
objects.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ! IMPORTANT if you save 200K+ cells</span></span>
<span><span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu">setAutoBlockSize</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1e+09</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html" class="external-link">saveHDF5SummarizedExperiment</a></span><span class="op">(</span></span>
<span>    <span class="st">"single_cell_counts"</span>, </span>
<span>    replace <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    as.sparse <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="saving-as-h5ad-slow-saving-fast-reading">Saving as H5AD (slow saving, fast reading)<a class="anchor" aria-label="anchor" href="#saving-as-h5ad-slow-saving-fast-reading"></a>
</h4>
<p>Saving as <code>.h5ad</code> executes any computation on the
<code>SingleCellExperiment</code> and writes it to disk as a monolithic
<code>H5AD</code>. The <code>H5AD</code> format is the HDF5 disk
representation of the AnnData object and is well-supported in
Python.</p>
<p>However this <code>.h5ad</code> saving strategy has a bottleneck of
handling columns with only NA values of a
<code>SingleCellExperiment</code> metadata.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ! IMPORTANT if you save 200K+ cells</span></span>
<span><span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu">setAutoBlockSize</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1e+09</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> <span class="fu">zellkonverter</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/writeH5AD.html" class="external-link">writeH5AD</a></span><span class="op">(</span><span class="st">"single_cell_counts.h5ad"</span>, </span>
<span>                                               compression <span class="op">=</span> <span class="st">"gzip"</span>,</span>
<span>                                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualise-gene-transcription">Visualise gene transcription<a class="anchor" aria-label="anchor" href="#visualise-gene-transcription"></a>
</h3>
<p>We can gather all CD14 monocytes cells and plot the distribution of
ENSG00000085265 (FCN1) across all tissues</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plots with styling</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># Filter and subset</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cell_type_unified_ensemble</span> <span class="op">==</span> <span class="st">"cd14 mono"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Get counts per million for FCN1 gene</span></span>
<span>  <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span>, features <span class="op">=</span> <span class="st">"ENSG00000085265"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Add feature to table</span></span>
<span>  <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu">join_features</span><span class="op">(</span><span class="st">"ENSG00000085265"</span>, shape <span class="op">=</span> <span class="st">"wide"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Rank x axis</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Rename to gene symbol</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>FCN1 <span class="op">=</span> <span class="va">ENSG00000085265</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot by disease</span></span>
<span><span class="va">counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/with_groups.html" class="external-link">with_groups</a></span><span class="op">(</span><span class="va">disease</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">`FCN1`</span>, rm.na<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html" class="external-link">fct_reorder</a></span><span class="op">(</span><span class="va">disease</span>, <span class="va">median_count</span>,.desc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">`FCN1`</span>,color <span class="op">=</span> <span class="va">dataset_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>shape<span class="op">=</span><span class="st">"."</span><span class="op">)</span> <span class="op">+</span></span>
<span>    </span>
<span>  <span class="co"># Style</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">60</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Disease"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"FCN1 in CD14 monocytes by disease. Coloured by datasets"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="../reference/figures/FCN1_disease_plot.png" width="434"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot by tissue</span></span>
<span><span class="va">counts</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/with_groups.html" class="external-link">with_groups</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">`FCN1`</span>, rm.na<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html" class="external-link">fct_reorder</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="va">median_count</span>,.desc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">`FCN1`</span>,color <span class="op">=</span> <span class="va">dataset_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>shape<span class="op">=</span><span class="st">"."</span><span class="op">)</span> <span class="op">+</span></span>
<span>    </span>
<span>  <span class="co"># Style</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">60</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Tissue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"FCN1 in CD14 monocytes by tissue. Colored by datasets"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="../reference/figures/FCN1_tissue_plot.png" width="1200"></p>
</div>
<div class="section level3">
<h3 id="integrate-cloud-and-local-metadata">Integrate cloud and local metadata<a class="anchor" aria-label="anchor" href="#integrate-cloud-and-local-metadata"></a>
</h3>
<p><code>cellNexus</code> not only enables users to query our metadata
but also allows integration with your local metadata. Additionally,
users can integrate with your metadata stored in the cloud.</p>
<p>To enable this feature, users must include
<code>file_id_cellNexus_single_cell</code> and <code>atlas_id</code>
(e.g cellxgene/dd-mm-yy) columns in the metadata. See metadata structure
in cellNexus::pbmc3k_sce</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up local cache and paths</span></span>
<span><span class="va">local_cache</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">layer</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span><span class="va">meta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">local_cache</span>, <span class="st">"pbmc3k_metadata.parquet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract and prepare metadata</span></span>
<span><span class="va">pbmc3k_metadata</span> <span class="op">&lt;-</span> <span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc3k_sce.html">pbmc3k_sce</a></span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    counts_directory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">atlas_id</span>, <span class="va">layer</span><span class="op">)</span>,</span>
<span>    sce_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">counts_directory</span>, <span class="va">file_id_cellNexus_single_cell</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Get unique paths</span></span>
<span><span class="va">counts_directory</span> <span class="op">&lt;-</span> <span class="va">pbmc3k_metadata</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">counts_directory</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_path</span> <span class="op">&lt;-</span> <span class="va">pbmc3k_metadata</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">sce_path</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create directory structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">counts_directory</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save data to disk</span></span>
<span><span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc3k_sce.html">pbmc3k_sce</a></span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/write_parquet.html" class="external-link">write_parquet</a></span><span class="op">(</span><span class="va">meta_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save SCE object</span></span>
<span><span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc3k_sce.html">pbmc3k_sce</a></span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">zellkonverter</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/writeH5AD.html" class="external-link">writeH5AD</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">sce_path</span>, compression <span class="op">=</span> <span class="st">"gzip"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Using the <span style="color: #00BB00;">'counts'</span> assay as the <span style="color: #00BB00;">X matrix</span></span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A cellNexus file</span></span>
<span><span class="va">file_id_from_cloud</span> <span class="op">&lt;-</span> <span class="st">"e52795dec7b626b6276b867d55328d9f___1.h5ad"</span></span>
<span><span class="va">file_id_local</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">sce_path</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cloud_metadata <span class="op">=</span> <span class="va">METADATA_URL</span>,</span>
<span>             local_metadata <span class="op">=</span> <span class="va">meta_path</span>,</span>
<span>             cache_directory <span class="op">=</span> <span class="va">local_cache</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># For illustration purpose, only filter a selected cloud metadata and the saved metadata </span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">file_id_cellNexus_single_cell</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">file_id_from_cloud</span>, <span class="va">file_id_local</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_id</span>, <span class="va">sample_id</span>, <span class="va">dataset_id</span>, <span class="va">cell_type_unified_ensemble</span>, <span class="va">atlas_id</span>, <span class="va">file_id_cellNexus_single_cell</span> <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span>cache_directory <span class="op">=</span> <span class="va">local_cache</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Downloading 1 file, totalling 0.02 GB</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Downloading https://object-store.rc.nectar.org.au/v1/AUTH_06d6e008e3e642da99d806ba3ea629c5/cellNexus-anndata/cellxgene/21-08-2025/counts/e52795dec7b626b6276b867d55328d9f___1.h5ad to /tmp/RtmprGqjm3/cellxgene/21-08-2025//counts/e52795dec7b626b6276b867d55328d9f___1.h5ad</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> cellNexus says: Not all genes completely overlap across the provided objects.Counts are generated by genes intersection.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 12795 3572 </span></span>
<span><span class="co">#&gt; metadata(1): data</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(12795): ENSG00000228463 ENSG00000228327 ... ENSG00000273748</span></span>
<span><span class="co">#&gt;   ENSG00000278384</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(3572): AAACATACAACCAC_1 AAACATTGAGCTAC_1 ...</span></span>
<span><span class="co">#&gt;   GCTGCGACACTAGTAC_5_liao___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span><span class="co">#&gt;   CTCACACCAGCTTAAC_5_liao___9f222629-9e39-47d0-b83f-e08d610c7479_2</span></span>
<span><span class="co">#&gt; colData names(7): sample_id dataset_id ... dir_prefix original_cell_</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="cell-metadata">Cell metadata<a class="anchor" aria-label="anchor" href="#cell-metadata"></a>
</h2>
<p>Dataset-specific columns (definitions available at
cellxgene.cziscience.com)</p>
<p><code>cell_count</code>, <code>collection_id</code>,
<code>filetype</code>, <code>is_primary_data</code>,
<code>mean_genes_per_cell</code>, <code>published_at</code>,
<code>revised_at</code>, <code>schema_version</code>,
<code>tombstone</code>, <code>x_normalization</code>,
<code>explorer_url</code>, <code>dataset_id</code>,
<code>dataset_version_id</code></p>
<p>Sample-specific columns (definitions available at
cellxgene.cziscience.com)</p>
<p><code>sample_id</code>, <code>sample_</code>, <code>age_days</code>,
<code>assay</code>, <code>assay_ontology_term_id</code>,
<code>development_stage</code>,
<code>development_stage_ontology_term_id</code>,
<code>self_reported_ethnicity</code>,
<code>self_reported_ethnicity_ontology_term_id</code>,
<code>experiment___</code>, <code>organism</code>,
<code>organism_ontology_term_id</code>, <code>sample_placeholder</code>,
<code>sex</code>, <code>sex_ontology_term_id</code>,
<code>tissue</code>, <code>tissue_type</code>,
<code>tissue_ontology_term_id</code>, <code>tissue_groups</code>,
<code>disease</code>, <code>disease_ontology_term_id</code>,
<code>is_primary_data</code>, <code>donor_id</code>,
<code>is_immune</code></p>
<p>Cell-specific columns (definitions available at
cellxgene.cziscience.com)</p>
<p><code>cell_id</code>, <code>cell_type</code>,
<code>cell_type_ontology_term_id</code>,
<code>cell_annotation_azimuth_l2</code>,
<code>cell_annotation_blueprint_singler</code>,
<code>observation_joinid</code>, <code>empty_droplet</code>,
<code>alive</code>, <code>scDblFinder.class</code></p>
<p>Through harmonisation and curation we introduced custom column, not
present in the original CELLxGENE metadata</p>
<ul>
<li>
<code>age_days</code>: donors’ age in days</li>
<li>
<code>cell_type_unified_ensemble</code>: the consensus call identity
(for immune cells) using the original and three novel annotations using
Seurat Azimuth and SingleR</li>
<li>
<code>cell_annotation_azimuth_l2</code>: Azimuth cell
annotation</li>
<li>
<code>cell_annotation_blueprint_singler</code>: SingleR cell
annotation using Blueprint reference</li>
<li>
<code>cell_annotation_blueprint_monaco</code>: SingleR cell
annotation using Monaco reference</li>
<li>
<code>sample_heuristic</code>: sample subdivision for internal
use</li>
<li>
<code>file_id_cellNexus_single_cell</code>: file subdivision for
internal use</li>
<li>
<code>file_id_cellNexus_pseudobulk</code>: file subdivision for
internal use</li>
<li>
<code>sample_id</code>: sample ID</li>
<li>
<code>nCount_RNA</code>: total number of RNA detected in a cell per
sample</li>
<li>
<code>nFeature_expressed_in_sample</code>: total number of genes
expressed in a cell per sample</li>
</ul>
</div>
<div class="section level2">
<h2 id="rna-abundance">RNA abundance<a class="anchor" aria-label="anchor" href="#rna-abundance"></a>
</h2>
<p>The <code>counts</code> assay includes RNA abundance in the positive
real scale (not transformed with non-linear functions, e.g. log sqrt).
Originally CELLxGENE include a mix of scales and transformations
specified in the <code>x_normalization</code> column.</p>
<p>The <code>cpm</code> assay includes counts per million.</p>
</div>
<div class="section level2">
<h2 id="other-representations">Other representations<a class="anchor" aria-label="anchor" href="#other-representations"></a>
</h2>
<p>The <code>rank</code> assay is the representation of each cell’s gene
expression profile where genes are ranked by expression intensity.</p>
<p>The <code>pseudobulk</code> assay includes aggregated RNA abundance
for sample and cell type combination.</p>
<p>The metacell (e.g <code>metacell_2</code>, <code>metacell_4</code>
etc) assays represent hierarchical partitions of cells into metacell
groups.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_4.0.0    cellNexus_0.99.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22            splines_4.5.1              </span></span>
<span><span class="co">#&gt;   [3] later_1.4.4                 filelock_1.0.3             </span></span>
<span><span class="co">#&gt;   [5] tibble_3.3.0                polyclip_1.10-7            </span></span>
<span><span class="co">#&gt;   [7] fastDummies_1.7.5           lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;   [9] rprojroot_2.1.1             globals_0.18.0             </span></span>
<span><span class="co">#&gt;  [11] lattice_0.22-7              MASS_7.3-65                </span></span>
<span><span class="co">#&gt;  [13] backports_1.5.0             magrittr_2.0.4             </span></span>
<span><span class="co">#&gt;  [15] plotly_4.11.0               sass_0.4.10                </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.30              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [19] yaml_2.3.10                 httpuv_1.6.16              </span></span>
<span><span class="co">#&gt;  [21] Seurat_5.3.0                sctransform_0.4.2          </span></span>
<span><span class="co">#&gt;  [23] spam_2.11-1                 sp_2.2-0                   </span></span>
<span><span class="co">#&gt;  [25] spatstat.sparse_3.1-0       reticulate_1.43.0          </span></span>
<span><span class="co">#&gt;  [27] cowplot_1.2.0               pbapply_1.7-4              </span></span>
<span><span class="co">#&gt;  [29] DBI_1.2.3                   RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [31] abind_1.4-8                 Rtsne_0.17                 </span></span>
<span><span class="co">#&gt;  [33] GenomicRanges_1.61.5        purrr_1.1.0                </span></span>
<span><span class="co">#&gt;  [35] BiocGenerics_0.55.1         IRanges_2.43.5             </span></span>
<span><span class="co">#&gt;  [37] S4Vectors_0.47.4            ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [39] irlba_2.3.5.1               listenv_0.9.1              </span></span>
<span><span class="co">#&gt;  [41] spatstat.utils_3.2-0        goftest_1.2-3              </span></span>
<span><span class="co">#&gt;  [43] RSpectra_0.16-2             spatstat.random_3.4-2      </span></span>
<span><span class="co">#&gt;  [45] fitdistrplus_1.2-4          parallelly_1.45.1          </span></span>
<span><span class="co">#&gt;  [47] pkgdown_2.1.3               codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [49] DelayedArray_0.35.3         tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [51] farver_2.1.2                shinyWidgets_0.9.0         </span></span>
<span><span class="co">#&gt;  [53] matrixStats_1.5.0           stats4_4.5.1               </span></span>
<span><span class="co">#&gt;  [55] spatstat.explore_3.5-3      Seqinfo_0.99.2             </span></span>
<span><span class="co">#&gt;  [57] duckdb_1.4.0                jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt;  [59] progressr_0.16.0            ggridges_0.5.7             </span></span>
<span><span class="co">#&gt;  [61] survival_3.8-3              systemfonts_1.2.3          </span></span>
<span><span class="co">#&gt;  [63] tools_4.5.1                 ragg_1.5.0                 </span></span>
<span><span class="co">#&gt;  [65] ica_1.0-3                   Rcpp_1.1.0                 </span></span>
<span><span class="co">#&gt;  [67] glue_1.8.0                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [69] SparseArray_1.9.1           xfun_0.53                  </span></span>
<span><span class="co">#&gt;  [71] MatrixGenerics_1.21.0       dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [73] HDF5Array_1.37.0            withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [75] fastmap_1.2.0               basilisk_1.21.5            </span></span>
<span><span class="co">#&gt;  [77] rhdf5filters_1.21.0         digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [79] R6_2.6.1                    mime_0.13                  </span></span>
<span><span class="co">#&gt;  [81] textshaping_1.0.3           scattermore_1.2            </span></span>
<span><span class="co">#&gt;  [83] tensor_1.5.1                spatstat.data_3.1-8        </span></span>
<span><span class="co">#&gt;  [85] h5mread_1.1.1               utf8_1.2.6                 </span></span>
<span><span class="co">#&gt;  [87] tidyr_1.3.1                 generics_0.1.4             </span></span>
<span><span class="co">#&gt;  [89] data.table_1.17.8           httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [91] htmlwidgets_1.6.4           S4Arrays_1.9.1             </span></span>
<span><span class="co">#&gt;  [93] uwot_0.2.3                  pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [95] gtable_0.3.6                blob_1.2.4                 </span></span>
<span><span class="co">#&gt;  [97] lmtest_0.9-40               S7_0.2.0                   </span></span>
<span><span class="co">#&gt;  [99] SingleCellExperiment_1.31.1 XVector_0.49.1             </span></span>
<span><span class="co">#&gt; [101] htmltools_0.5.8.1           dotCall64_1.2              </span></span>
<span><span class="co">#&gt; [103] SeuratObject_5.2.0          scales_1.4.0               </span></span>
<span><span class="co">#&gt; [105] Biobase_2.69.1              png_0.1-8                  </span></span>
<span><span class="co">#&gt; [107] spatstat.univar_3.1-4       knitr_1.50                 </span></span>
<span><span class="co">#&gt; [109] reshape2_1.4.4              checkmate_2.3.3            </span></span>
<span><span class="co">#&gt; [111] nlme_3.1-168                curl_7.0.0                 </span></span>
<span><span class="co">#&gt; [113] cachem_1.1.0                zoo_1.8-14                 </span></span>
<span><span class="co">#&gt; [115] rhdf5_2.53.5                stringr_1.5.2              </span></span>
<span><span class="co">#&gt; [117] KernSmooth_2.23-26          parallel_4.5.1             </span></span>
<span><span class="co">#&gt; [119] miniUI_0.1.2                zellkonverter_1.19.2       </span></span>
<span><span class="co">#&gt; [121] arrow_21.0.0.1              desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [123] pillar_1.11.1               grid_4.5.1                 </span></span>
<span><span class="co">#&gt; [125] vctrs_0.6.5                 RANN_2.6.2                 </span></span>
<span><span class="co">#&gt; [127] promises_1.3.3              dbplyr_2.5.1               </span></span>
<span><span class="co">#&gt; [129] xtable_1.8-4                cluster_2.1.8.1            </span></span>
<span><span class="co">#&gt; [131] evaluate_1.0.5              cli_3.6.5                  </span></span>
<span><span class="co">#&gt; [133] compiler_4.5.1              rlang_1.1.6                </span></span>
<span><span class="co">#&gt; [135] crayon_1.5.3                future.apply_1.20.0        </span></span>
<span><span class="co">#&gt; [137] plyr_1.8.9                  fs_1.6.6                   </span></span>
<span><span class="co">#&gt; [139] stringi_1.8.7               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [141] deldir_2.0-4                assertthat_0.2.1           </span></span>
<span><span class="co">#&gt; [143] lazyeval_0.2.2              spatstat.geom_3.6-0        </span></span>
<span><span class="co">#&gt; [145] Matrix_1.7-4                dir.expiry_1.17.0          </span></span>
<span><span class="co">#&gt; [147] RcppHNSW_0.6.0              patchwork_1.3.2            </span></span>
<span><span class="co">#&gt; [149] bit64_4.6.0-1               future_1.67.0              </span></span>
<span><span class="co">#&gt; [151] Rhdf5lib_1.31.0             shiny_1.11.1               </span></span>
<span><span class="co">#&gt; [153] SummarizedExperiment_1.39.2 ROCR_1.0-11                </span></span>
<span><span class="co">#&gt; [155] igraph_2.1.4                bslib_0.9.0                </span></span>
<span><span class="co">#&gt; [157] bit_4.6.0</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola, Mengyuan Shen, Michael Milton, Silicon Valley Foundation CZF2019-002443, NIH NHGRI 5U24HG004059-18, Victoria Cancer Agency ECRF21036, NHMRC 1116955.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
